keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,531 @@
+However, the user of the namespace won't notice why it is blocked and won't find a way to delete the secrets other than manually deleting finalizer.
+In addition, to make things worse, provisioner secrets that are defined by using template with PVC information,
+like "${pvc.name}", would be deleted in above case, due to the failure in finding a reference to the secret,
+because the reference to provisioner secret is not stored in PV and needs to be resolved on deletion time with PVC information that is already deleted.
+This behavior is inconsistent with other secrets, like node publish secret whose reference is stored in PV.
+
+## Design Details
+
+This feature can be implemented in the same way as `pv-protection/pvc-protection`.
+Protection logics for in-tree resources and out-of-tree resources are separeted and independently work.
+It is due to the restriction that CRDs can't be handled from in-tree controller. Note that in-tree controller protect both Secrets and ConfigMaps, on the other hand, out-of-tree controller protect only Secrets because VoluemSnapshots never use ConfigMaps.
+
+- In-tree resources(`Pod` and `PersistentVolume`):
+  - The deletion is blocked by using newly introduced `kubernetes.io/ephemeral-protection` finalizer,
+  - The `kubernetes.io/ephemeral-protection` finalizer will be always added on creation of the Secret/ConfigMap by using admission controller for in-tree resources,
+  - After the Secret/ConfigMap is requested to be deleted (`deletionTimestamp` is set), the `kubernetes.io/ephemeral-protection` finalizer will be deleted by newly introduced in-tree `ephemeral-protection-controller` by checking whether the Secret/ConfigMap is in-use, on every change(Create/Update/Delete) events for Secrets/ConfigMaps and related resources.
+  - If the `EphemeralInUseProtection` feature gate is disalbed, finalizer is added on creation, but deleted regardless of whether it is in-use after `deletionTimestamp` is set. This will allow users to avoid manually deleting finalizers on the downgrading by disabling the feature.
+- Out-of-tree resources(`VolumeSnapshot`):
+  - The deletion is blocked by using newly introduced `snapshot.storage.kubernetes.io/ephemeral-protection` finalizer,
+  - The `snapshot.storage.kubernetes.io/ephemeral-protection` finalizer will be always added on creation of the secret by using snapshot controller,
+  - After the secret is requested to be deleted (`deletionTimestamp` is set), the `snapshot.storage.kubernetes.io/ephemeral-protection` finalizer will be deleted by newly introduced out-of-tree `ephemeral-protection-volumesnapshot-controller` by checking whether the secret is in-use, on every change(Create/Update/Delete) events for secrets and related resources.
+
+Feature gate `EphemeralInUseProtection` is used only for in-tree controller. Out-of-tree controller will be enabled when the `ephemeral-protection-volumesnapshot-controller` is deployed with `--ephemeral-protection=enabled` flag.
```

Comment:
This will be one of the choices, however what will be the actual cases saved by implementing this way?  Isn't it just complicating things?

Reply:
That seems required to safely roll out finalizer enablement in a way that doesn't hang the resources if the controller deployment doesn't complete or has to be rolled back

Reply:
I may miss the point, but if we consider the case that controller deployment doesn't complete, adding flags won't help avoiding hangs of the resource, because no controller means no way to remove the finalizer by it.

With a bit of more consideration, I found that adding this flag will be useful when users really want to delete the out-of-tree controller completely. In this case, disabling finalizer-adding logic helps to ensure that no one add new finalizers while they are removing finalizers manually. Without it, there will be a case that the finalizer is added before they disable finalizer-removing.

However, in most cases, users won't require removing out-of-tree controller completely, if they have a way to disable the feature. Even if they need it, they can just remove finalizer before and after removing out-of-tree controller. There may be a small amount of time that some secrets aren't deleted between out-of-tree controller is removed and finalizers are removed, however it won't be an issue because the secrets will eventually be deleted. 

Any other cases or does the case really matter?